Switch `SideEffect` to GAT lifetime

Side effects, when expressed correctly, should just be functions that work for any given `SideEffectRegistrar<'a>` and should themselves not be dependent upon any lifetime (i.e., the `SideEffectRegistrar` itself). However, [due to limitations in Rust's current implementation of higher-ranked lifetimes](https://github.com/rust-lang/rust/issues/116869) and some other related weird type-related bugs I've found along the way, this is not possible at the moment and the current (technically incorrect) implementation of side effects was born (they work, just not optimally). Rust just isn't quite ready for some of the odd, edge-case stuff I am trying to put it through.

Here's some sample code I was working on that properly expresses side effects, for when the time comes that it is possible:
```rust
pub trait SideEffect {
    type Api<'a>;
    fn build<'a>(self, registrar: SideEffectRegistrar<'a>) -> Self::Api<'a>;
}
impl<T, F: FnOnce(SideEffectRegistrar) -> T> SideEffect for F {
    type Api<'a> = T;
    fn build<'a>(self, registrar: SideEffectRegistrar<'a>) -> Self::Api<'a> {
        self(registrar)
    }
}
```

And for future reference, if I need it, here's some side-effect code I was experimenting with using the correct approach.

```rust
// pub fn raw<T: Send + 'static>(
//     initial: T,
// ) -> impl for<'a> SideEffect<
//     Api<'a> = (
//         &'a mut T,
//         impl Fn(Box<dyn FnOnce(&mut T)>) + Clone + Send + Sync,
//     ),
// > {
//     fix_lifetime(move |register: SideEffectRegistrar| register.raw(initial))
// }
// pub fn raw<T: Send + 'static>(
//     initial: T,
// ) -> impl for<'a> FnOnce(
//     SideEffectRegistrar<'a>,
// ) -> (
//     &'a mut T,
//     impl Fn(Box<dyn FnOnce(&mut T)>) + Clone + Send + Sync,
// ) {
//     fix_lifetime(move |register: SideEffectRegistrar| register.raw(initial))
// }

fn fix_lifetime<F, T, R>(f: F) -> F
where
    F: for<'a> FnOnce(SideEffectRegistrar<'a>) -> (&'a mut T, R),
{
    f
}

pub fn as_listener() -> impl for<'a> SideEffect<Api<'a> = ()> {
    |_: SideEffectRegistrar| {}
}

// TODO THis works because it's not a closure and the lifetime elision rules work correctly
pub fn zero_raw(register: SideEffectRegistrar) -> (&mut u8, impl Fn(u8) + Clone + Send + Sync) {
    let (state, rebuild) = register.raw(0);
    let set_state = move |new_state| {
        rebuild(Box::new(move |state| *state = new_state));
    };
    (state, set_state)
}
```

Tried the following as a workaround:

```rust
pub struct Raw<T>(T);
impl<T: Send + 'static> SideEffect for Raw<T> {
    type Api<'a> = (&'a mut T, impl CData + Fn(Box<dyn FnOnce(&mut T)>));
    fn build(self, registrar: SideEffectRegistrar<'_>) -> Self::Api<'_> {
        registrar.raw(self.0)
    }
}

pub fn as_listener() -> impl for<'a> SideEffect<Api<'a> = ()> {
    |_: SideEffectRegistrar| {}
}

pub struct State<T>(pub T);
impl<T: Send + 'static> SideEffect for State<T> {
    type Api<'a> = (&'a mut T, impl CData + Fn(T));
    fn build(self, registrar: SideEffectRegistrar<'_>) -> Self::Api<'_> {
        let (state, rebuild) = registrar.raw(self.0);
        let set_state = move |new_state| {
            rebuild(Box::new(|state| *state = new_state));
        };
        (state, set_state)
    }
}
```

Didn't work because:
- A slew of compiler bugs. Don't even know where to begin
  - Big one is several ```error[E0700]: hidden type for `impl (Fn(u8)) + CData` captures lifetime that does not appear in bounds``` and related messages, but that makes no sense because the `impl Trait` is `'static` and doesn't depend on the lifetime `'a` from the higher ranked lifetime (it does a `.clone()` of the needed elements to be `'static`).
- Requires nightly for `impl_trait_in_assoc_type`, which has been attempted to be pushed to stabilization, but is currently stalled.

So, really looks like this issue is impossible to achieve until Rust has some substantial compiler bug fixes.

--- a/rearch-effects/src/lib.rs
+++ b/rearch-effects/src/lib.rs
@@ -1,43 +1,114 @@
 use rearch::{CData, SideEffect, SideEffectRegistrar};
 use std::{cell::OnceCell, sync::Arc};
 
-pub fn raw<'a, T: Send + 'static>(
+// This workaround was derived from:
+// https://github.com/GregoryConrad/rearch-rs/issues/3#issuecomment-1872869363
+// And is needed because of:
+// https://github.com/rust-lang/rust/issues/111662
+use effect_lifetime_fixers::{EffectLifetimeFixer0, EffectLifetimeFixer1, EffectLifetimeFixer2};
+mod effect_lifetime_fixers {
+    use rearch::{SideEffect, SideEffectRegistrar};
+
+    pub struct EffectLifetimeFixer0<F>(F);
+    impl<T, F> SideEffect for EffectLifetimeFixer0<F>
+    where
+        T: Send + 'static,
+        F: FnOnce(SideEffectRegistrar) -> &mut T,
+    {
+        type Api<'a> = &'a mut T;
+        fn build(self, registrar: SideEffectRegistrar) -> Self::Api<'_> {
+            self.0(registrar)
+        }
+    }
+    impl<F> EffectLifetimeFixer0<F> {
+        pub(super) const fn new<T>(f: F) -> Self
+        where
+            F: FnOnce(SideEffectRegistrar) -> &mut T,
+        {
+            Self(f)
+        }
+    }
+
+    pub struct EffectLifetimeFixer1<F>(F);
+    impl<T, F, R1> SideEffect for EffectLifetimeFixer1<F>
+    where
+        T: Send + 'static,
+        F: FnOnce(SideEffectRegistrar) -> (&mut T, R1),
+    {
+        type Api<'a> = (&'a mut T, R1);
+        fn build(self, registrar: SideEffectRegistrar) -> Self::Api<'_> {
+            self.0(registrar)
+        }
+    }
+    impl<F> EffectLifetimeFixer1<F> {
+        pub(super) const fn new<T, R1>(f: F) -> Self
+        where
+            F: FnOnce(SideEffectRegistrar) -> (&mut T, R1),
+        {
+            Self(f)
+        }
+    }
+
+    pub struct EffectLifetimeFixer2<F>(F);
+    impl<T, F, R1, R2> SideEffect for EffectLifetimeFixer2<F>
+    where
+        T: Send + 'static,
+        F: FnOnce(SideEffectRegistrar) -> (&mut T, R1, R2),
+    {
+        type Api<'a> = (&'a mut T, R1, R2);
+        fn build(self, registrar: SideEffectRegistrar) -> Self::Api<'_> {
+            self.0(registrar)
+        }
+    }
+    impl<F> EffectLifetimeFixer2<F> {
+        pub(super) const fn new<T, R1, R2>(f: F) -> Self
+        where
+            F: FnOnce(SideEffectRegistrar) -> (&mut T, R1, R2),
+        {
+            Self(f)
+        }
+    }
+}
+
+pub fn raw<T: Send + 'static>(
     initial: T,
-) -> impl SideEffect<
-    'a,
-    Api = (
+) -> impl for<'a> SideEffect<
+    Api<'a> = (
         &'a mut T,
         impl CData + Fn(Box<dyn FnOnce(&mut T)>),
         Arc<dyn Send + Sync + Fn(Box<dyn FnOnce()>)>,
     ),
 > {
-    move |register: SideEffectRegistrar<'a>| register.raw(initial)
+    EffectLifetimeFixer2::new(move |register: SideEffectRegistrar| register.raw(initial))
 }
 
+// NOTE: returns (), the no-op side effect
 #[must_use]
-pub fn as_listener<'a>() -> impl SideEffect<'a, Api = ()> {} // returns (), the no-op side effect
+pub fn as_listener() -> impl for<'a> SideEffect<Api<'a> = ()> {}
 
-pub fn state<'a, T: Send + 'static>(
+pub fn state<T: Send + 'static>(
     initial: T,
-) -> impl SideEffect<'a, Api = (&'a mut T, impl CData + Fn(T))> {
-    move |register: SideEffectRegistrar<'a>| {
+) -> impl for<'a> SideEffect<Api<'a> = (&'a mut T, impl CData + Fn(T))> {
+    EffectLifetimeFixer1::new(move |register: SideEffectRegistrar| {
         let (state, rebuild, _) = register.raw(initial);
         let set_state = move |new_state| {
             rebuild(Box::new(|state| *state = new_state));
         };
         (state, set_state)
-    }
+    })
 }
 
 // This uses a hacked together Lazy implementation because LazyCell doesn't have force_mut;
 // see https://github.com/rust-lang/rust/issues/109736#issuecomment-1605787094
 #[allow(clippy::missing_panics_doc)] // false positive
-pub fn lazy_state<'a, T, F>(init: F) -> impl SideEffect<'a, Api = (&'a mut T, impl CData + Fn(T))>
+pub fn lazy_state<T, F>(
+    init: F,
+) -> impl for<'a> SideEffect<Api<'a> = (&'a mut T, impl CData + Fn(T))>
 where
     T: Send + 'static,
     F: FnOnce() -> T + Send + 'static,
 {
-    move |register: SideEffectRegistrar<'a>| {
+    EffectLifetimeFixer1::new(move |register: SideEffectRegistrar| {
         let ((cell, f), rebuild, _) = register.raw((OnceCell::new(), Some(init)));
         cell.get_or_init(|| std::mem::take(f).expect("Init fn should be present for cell init")());
         let state = cell.get_mut().expect("State initialized above");
@@ -48,34 +119,31 @@ where
             }));
         };
         (state, set_state)
-    }
+    })
 }
 
-pub fn value<'a, T: Send + 'static>(value: T) -> impl SideEffect<'a, Api = &'a mut T> {
-    move |register: SideEffectRegistrar<'a>| {
-        let (state, _, _) = register.raw(value);
-        state
-    }
+pub fn value<T: Send + 'static>(value: T) -> impl for<'a> SideEffect<Api<'a> = &'a mut T> {
+    EffectLifetimeFixer0::new(move |register: SideEffectRegistrar| register.raw(value).0)
 }
 
 // This uses a hacked together Lazy implementation because LazyCell doesn't have force_mut;
 // see https://github.com/rust-lang/rust/issues/109736#issuecomment-1605787094
 #[allow(clippy::missing_panics_doc)] // false positive
-pub fn lazy_value<'a, T, F>(init: F) -> impl SideEffect<'a, Api = &'a mut T>
+pub fn lazy_value<T, F>(init: F) -> impl for<'a> SideEffect<Api<'a> = &'a mut T>
 where
     T: Send + 'static,
     F: FnOnce() -> T + Send + 'static,
 {
-    move |register: SideEffectRegistrar<'a>| {
+    EffectLifetimeFixer0::new(move |register: SideEffectRegistrar| {
         let ((cell, f), _, _) = register.raw((OnceCell::new(), Some(init)));
         cell.get_or_init(|| std::mem::take(f).expect("Init fn should be present for cell init")());
         cell.get_mut().expect("State initialized above")
-    }
+    })
 }
 
 #[must_use]
-pub fn is_first_build<'a>() -> impl SideEffect<'a, Api = bool> {
-    move |register: SideEffectRegistrar<'a>| {
+pub fn is_first_build() -> impl for<'a> SideEffect<Api<'a> = bool> {
+    move |register: SideEffectRegistrar| {
         let has_built_before = register.register(value(false));
         let is_first_build = !*has_built_before;
         *has_built_before = true;
@@ -88,30 +156,22 @@ pub fn is_first_build<'a>() -> impl SideEffect<'a, Api = bool> {
 /// Only use this side effect when you don't have any state that you wish to update in the rebuild
 /// (which is extremely rare).
 #[must_use]
-pub fn rebuilder<'a>() -> impl SideEffect<'a, Api = impl CData + Fn()> {
-    move |register: SideEffectRegistrar<'a>| {
+pub fn rebuilder() -> impl for<'a> SideEffect<Api<'a> = impl CData + Fn()> {
+    move |register: SideEffectRegistrar| {
         let ((), rebuild, _) = register.raw(());
         move || rebuild(Box::new(|()| {}))
     }
 }
 
-pub fn run_once<'a, F>(f: F) -> impl SideEffect<'a, Api = ()>
-where
-    F: FnOnce() + Send + 'static,
-{
-    move |register: SideEffectRegistrar<'a>| {
-        register.register(lazy_value(f));
-    }
-}
-
+/*
 /// Side effect that runs a callback whenever it changes and is dropped.
 /// Similar to `useEffect` from React.
 #[must_use]
-pub fn run_on_change<'a, F>() -> impl SideEffect<'a, Api = impl FnMut(F) + 'a>
+pub fn run_on_change<F>() -> impl for<'a> SideEffect<Api<'a> = impl FnMut(F) + 'a>
 where
     F: FnOnce() + Send + 'static,
 {
-    move |register: SideEffectRegistrar<'a>| {
+    move |register: SideEffectRegistrar| {
         let state = register.register(value(FunctionalDrop(None)));
         // The old callback, if there is one, will be called when it is dropped,
         // via the `*state = ...` assignment below
@@ -126,16 +186,34 @@ impl<F: FnOnce()> Drop for FunctionalDrop<F> {
         }
     }
 }
+#[must_use]
+pub fn run_on_change2<F>() -> RunOnChange<F>
+where
+    F: FnOnce() + Send + 'static,
+{
+    RunOnChange(std::marker::PhantomData)
+}
+pub struct RunOnChange<F>(std::marker::PhantomData<F>);
+impl<F: Send + FnOnce() + 'static> SideEffect for RunOnChange<F> {
+    type Api<'registrar> = impl FnMut(F) + 'registrar;
 
-pub fn reducer<'a, State, Action, Reducer>(
+    fn build(self, registrar: SideEffectRegistrar) -> Self::Api<'_> {
+        let state = registrar.register(value(FunctionalDrop(None)));
+        // The old callback, if there is one, will be called when it is dropped,
+        // via the `*state = ...` assignment below
+        |callback| *state = FunctionalDrop(Some(callback))
+    }
+}
+
+pub fn reducer<State, Action, Reducer>(
     reducer: Reducer,
     initial: State,
-) -> impl SideEffect<'a, Api = (State, impl CData + Fn(Action))>
+) -> impl for<'a> SideEffect<Api<'a> = (State, impl CData + Fn(Action))>
 where
     State: Clone + Send + Sync + 'static,
     Reducer: Fn(State, Action) -> State + Clone + Send + Sync + 'static,
 {
-    move |register: SideEffectRegistrar<'a>| {
+    move |register: SideEffectRegistrar<'_>| {
         let (state, set_state) = register.register(state(initial));
         (state.clone(), {
             let state = state.clone();
@@ -147,15 +225,15 @@ where
     }
 }
 
-pub fn lazy_reducer<'a, State, Action, Reducer>(
+pub fn lazy_reducer<State, Action, Reducer>(
     reducer: Reducer,
     initial: impl FnOnce() -> State + Send + 'static,
-) -> impl SideEffect<'a, Api = (State, impl CData + Fn(Action))>
+) -> impl for<'a> SideEffect<Api<'a> = (State, impl CData + Fn(Action))>
 where
     State: Clone + Send + Sync + 'static,
     Reducer: Fn(State, Action) -> State + Clone + Send + Sync + 'static,
 {
-    move |register: SideEffectRegistrar<'a>| {
+    move |register: SideEffectRegistrar| {
         let (state, set_state) = register.register(lazy_state(initial));
         (state.clone(), {
             let state = state.clone();
@@ -167,7 +245,7 @@ where
     }
 }
 
-/*
+// TODO should this actually be hydrate like in Dart?
 /// A thin wrapper around the state side effect that enables easy state persistence.
 ///
 /// You provide a `read` function and a `write` function,
@@ -177,17 +255,17 @@ where
 /// Note: when possible, it is highly recommended to use async persist instead of sync persist.
 /// This effect is blocking, which will prevent other capsule updates.
 /// However, this function is perfect for quick I/O, like when using something similar to redb.
-pub fn sync_persist<'a, Read, Write, R, T>(
+pub fn sync_persist<Read, Write, R, T>(
     read: Read,
     write: Write,
-) -> impl SideEffect<'a, Api = (&'a R, impl CData + Fn(T))>
+) -> impl for<'a> SideEffect<Api<'a> = (&'a R, impl CData + Fn(T))>
 where
     T: Send + 'static,
     R: Send + 'static,
     Read: FnOnce() -> R + Send + 'static,
     Write: Fn(T) -> R + Send + Sync + 'static,
 {
-    move |register: SideEffectRegistrar<'a>| {
+    move |register: SideEffectRegistrar| {
         let (state, set_state) = register.register(lazy_state(read));
         let write = Arc::new(write);
         let persist = move |new_data| {

--- a/rearch-macros/src/lib.rs
+++ b/rearch-macros/src/lib.rs
@@ -28,11 +28,11 @@ pub fn generate_tuple_side_effect_impl(input: TokenStream) -> TokenStream {
         }
     });
     let effect_impl = quote! {
-        impl<'a, #(#types: SideEffect<'a>),*> SideEffect<'a> for (#(#types),*) {
-            type Api = (#(#types::Api),*);
+        impl<#(#types: SideEffect),*> SideEffect for (#(#types),*) {
+            type Api<'registrar> = (#(#types::Api<'registrar>),*);
 
             #[allow(clippy::unused_unit)]
-            fn build(self, registrar: SideEffectRegistrar<'a>) -> Self::Api {
+            fn build(self, registrar: SideEffectRegistrar<'_>) -> Self::Api<'_> {
                 let (all_states, rebuild_all, run_txn) = registrar.raw((
                     #(#once_cell_inits),*
                 ));

--- a/rearch-tokio/src/lib.rs
+++ b/rearch-tokio/src/lib.rs
@@ -1,4 +1,4 @@
-use std::{cell::RefCell, future::Future, rc::Rc, sync::Arc};
+use std::future::Future;
 
 use rearch::{SideEffect, SideEffectRegistrar};
 use rearch_effects as effects;
@@ -43,9 +43,10 @@ impl<T> AsyncPersistState<T> {
     }
 }
 
+/*
 #[must_use]
-pub fn future<'a, T, F>(
-) -> impl SideEffect<'a, Api = (impl Fn() -> AsyncState<T> + 'a, impl FnMut(F) + 'a)>
+pub fn future<T, F>(
+) -> impl for<'a> SideEffect<Api<'a> = (impl Fn() -> AsyncState<T> + 'a, impl FnMut(F) + 'a)>
 where
     T: Clone + Send + 'static,
     F: Future<Output = T> + Send + 'static,
@@ -75,11 +76,11 @@ where
         (get, set)
     }
 }
+*/
 
 #[must_use]
-pub fn mutation<'a, T, F>() -> impl SideEffect<
-    'a,
-    Api = (
+pub fn mutation<T, F>() -> impl for<'a> SideEffect<
+    Api<'a> = (
         AsyncState<T>,
         impl Fn(F) + Clone + Send + Sync,
         impl Fn() + Clone + Send + Sync,
@@ -89,7 +90,7 @@ where
     T: Clone + Send + 'static,
     F: Future<Output = T> + Send + 'static,
 {
-    move |register: SideEffectRegistrar<'a>| {
+    move |register: SideEffectRegistrar| {
         let ((state, rebuild, _), (_, on_change)) = register.register((
             effects::raw(AsyncState::Idle(None)),
             // This immitates run_on_change, but for external use (outside of build)
@@ -125,10 +126,11 @@ where
     }
 }
 
-pub fn async_persist<'a, T, R, Reader, Writer, ReadFuture, WriteFuture>(
+/*
+pub fn async_persist<T, R, Reader, Writer, ReadFuture, WriteFuture>(
     read: Reader,
     write: Writer,
-) -> impl SideEffect<'a, Api = (AsyncPersistState<R>, impl FnMut(T) + Send + Sync + Clone)>
+) -> impl for<'a> SideEffect<Api<'a> = (AsyncPersistState<R>, impl FnMut(T) + Send + Sync + Clone)>
 where
     T: Send + 'static,
     R: Clone + Send + 'static,
@@ -137,7 +139,7 @@ where
     ReadFuture: Future<Output = R> + Send + 'static,
     WriteFuture: Future<Output = R> + Send + 'static,
 {
-    move |register: SideEffectRegistrar<'a>| {
+    move |register: SideEffectRegistrar| {
         let ((get_read, mut set_read), (write_state, set_write, _), is_first_build) =
             register.register((future(), mutation(), effects::is_first_build()));
 
@@ -167,3 +169,4 @@ where
         (state, persist)
     }
 }
+*/

--- a/rearch/src/lib.rs
+++ b/rearch/src/lib.rs
@@ -92,10 +92,7 @@ pub struct CapsuleHandle<'txn_scope, 'txn_total, 'build> {
 /// The key observation about side effects is that they form a tree, where each side effect:
 /// - Has its own private state (including composing other side effects together)
 /// - Presents some api to the build method, probably including a way to rebuild & update its state
-///
-/// *DO NOT MANUALLY IMPLEMENT THIS TRAIT YOURSELF!*
-/// It is an internal implementation detail that could be changed or removed in the future.
-pub trait SideEffect<'a> {
+pub trait SideEffect {
     /// The type exposed in the capsule build function when this side effect is registered;
     /// in other words, this is the api exposed by the side effect.
     ///
@@ -103,14 +100,14 @@ pub trait SideEffect<'a> {
     /// - Data and/or state in this side effect
     /// - Function callbacks (perhaps to trigger a rebuild and/or update the side effect state)
     /// - Anything else imaginable!
-    type Api;
+    type Api<'registrar>;
 
     /// Construct this side effect's `Api` via the given [`SideEffectRegistrar`].
-    fn build(self, registrar: SideEffectRegistrar<'a>) -> Self::Api;
+    fn build(self, registrar: SideEffectRegistrar) -> Self::Api<'_>;
 }
-impl<'a, T, F: FnOnce(SideEffectRegistrar<'a>) -> T> SideEffect<'a> for F {
-    type Api = T;
-    fn build(self, registrar: SideEffectRegistrar<'a>) -> Self::Api {
+impl<T, F: FnOnce(SideEffectRegistrar) -> T> SideEffect for F {
+    type Api<'registrar> = T;
+    fn build(self, registrar: SideEffectRegistrar) -> Self::Api<'_> {
         self(registrar)
     }
 }
@@ -185,7 +182,6 @@ impl Container {
         capsules.read(self)
     }
 
-    /* TODO(GregoryConrad): uncomment this listener section once we have side effects figured out
     /// Provides a mechanism to *temporarily* listen to changes in some capsule(s).
     /// The provided listener is called once at the time of the listener's registration,
     /// and then once again everytime a dependency changes.
@@ -195,7 +191,7 @@ impl Container {
     ///
     /// Thus, if you want the handle to live for as long as the Container itself,
     /// it is instead recommended to create a non-idempotent capsule
-    /// (use the [`side_effects::as_listener()`] side effect)
+    /// (use the `effects::as_listener()` side effect)
     /// that acts as your listener. When you normally would call `Container::listen()`,
     /// instead call `container.read(my_non_idempotent_listener)` to initialize it.
     ///
@@ -203,25 +199,24 @@ impl Container {
     /// Panics if you attempt to register the same listener twice,
     /// before the first `ListenerHandle` is dropped.
     #[must_use]
-    pub fn listen<ListenerEffect, EffectFactory, Listener>(
+    pub fn listen<Effect, EffectFactory, Listener>(
         &self,
         effect_factory: EffectFactory,
         listener: Listener,
     ) -> ListenerHandle
     where
-        ListenerEffect: for<'a> SideEffect<'a>,
-        EffectFactory: Fn() -> ListenerEffect + Send + Clone + 'static,
-        Listener: Fn(CapsuleReader, <ListenerEffect as SideEffect>::Api) + Send + 'static,
+        Effect: SideEffect,
+        EffectFactory: 'static + Send + Fn() -> Effect,
+        Listener: Fn(CapsuleReader, <Effect as SideEffect>::Api<'_>) + Send + 'static,
     {
         // We make a temporary non-idempotent capsule for the listener so that
         // it doesn't get disposed by the idempotent gc
         let tmp_capsule = move |CapsuleHandle { get, register }: CapsuleHandle| {
-            let effect_factory = effect_factory.clone();
             let effect = effect_factory();
-            let effect_state = register.register(effect);
-            listener(get, effect_state);
+            let effect_api = register.register(effect);
+            listener(get, effect_api);
         };
-        let id = tmp_capsule.type_id();
+        let id = tmp_capsule.id();
 
         // Put the temporary capsule into the container to listen to updates
         self.with_write_txn(move |txn| {
@@ -239,7 +234,6 @@ impl Container {
             store: Arc::downgrade(&self.0),
         }
     }
-    */
 }
 
 /// Represents a handle onto a particular listener, as created with `Container::listen()`.
@@ -259,7 +253,7 @@ pub struct ListenerHandle {
 impl Drop for ListenerHandle {
     fn drop(&mut self) {
         if let Some(store) = self.store.upgrade() {
-            // Note: The node is guaranteed to be in the graph here since it is a listener.
+            // NOTE: The node is guaranteed to be in the graph here since it is a listener.
             let orchestrator = SideEffectTxnOrchestrator(Weak::clone(&self.store));
             store.with_write_txn(orchestrator, |txn| txn.dispose_node(&self.id));
         }
@@ -509,26 +503,33 @@ mod tests {
     mod effects {
         use super::*;
 
-        pub fn as_listener<'a>() -> impl SideEffect<'a, Api = ()> {}
+        pub fn as_listener() -> impl for<'a> SideEffect<Api<'a> = ()> {}
 
-        pub fn state<'a, T: Send + 'static>(
+        pub fn cloned_state<T: Clone + Send + 'static>(
             initial: T,
-        ) -> impl SideEffect<'a, Api = (&'a mut T, impl CData + Fn(T))> {
-            move |register: SideEffectRegistrar<'a>| {
+        ) -> impl for<'a> SideEffect<Api<'a> = (T, impl CData + Fn(T))> {
+            move |register: SideEffectRegistrar| {
                 let (state, rebuild, _) = register.raw(initial);
                 let set_state = move |new_state| {
                     rebuild(Box::new(|state| *state = new_state));
                 };
-                (state, set_state)
+                (state.clone(), set_state)
             }
         }
 
-        pub fn is_first_build<'a>() -> impl SideEffect<'a, Api = bool> {
-            move |register: SideEffectRegistrar<'a>| {
+        pub fn is_first_build() -> impl for<'a> SideEffect<Api<'a> = bool> {
+            move |register: SideEffectRegistrar| {
                 let (has_built_before, _, _) = register.raw(false);
                 let is_first_build = !*has_built_before;
                 *has_built_before = true;
                 is_first_build
+            }
+        }
+
+        pub fn rebuilder() -> impl for<'a> SideEffect<Api<'a> = impl CData + Fn()> {
+            move |register: SideEffectRegistrar| {
+                let (_, rebuild, _) = register.raw(());
+                move || rebuild(Box::new(|_| {}))
             }
         }
     }
@@ -572,8 +573,7 @@ mod tests {
         use super::*;
 
         fn stateful(CapsuleHandle { register, .. }: CapsuleHandle) -> (u8, impl CData + Fn(u8)) {
-            let (state, set_state) = register.register(effects::state(0));
-            (*state, set_state)
+            register.register(effects::cloned_state(0))
         }
 
         fn dependent(CapsuleHandle { mut get, .. }: CapsuleHandle) -> u8 {
@@ -617,8 +617,9 @@ mod tests {
         fn foo(
             CapsuleHandle { register, .. }: CapsuleHandle,
         ) -> (u8, u8, impl CData + Fn(u8), impl CData + Fn(u8)) {
-            let ((s1, ss1), (s2, ss2)) = register.register((effects::state(0), effects::state(1)));
-            (*s1, *s2, ss1, ss2)
+            let ((s1, ss1), (s2, ss2)) =
+                register.register((effects::cloned_state(0), effects::cloned_state(1)));
+            (s1, s2, ss1, ss2)
         }
 
         let container = Container::new();
@@ -637,15 +638,8 @@ mod tests {
     #[cfg(feature = "better-api")]
     #[test]
     fn get_and_register() {
-        fn rebuilder<'a>() -> impl SideEffect<'a, Api = impl CData + Fn()> {
-            move |register: SideEffectRegistrar<'a>| {
-                let (_, rebuild, _) = register.raw(());
-                move || rebuild(Box::new(|_| {}))
-            }
-        }
-
         fn rebuildable(CapsuleHandle { register, .. }: CapsuleHandle) -> impl CData + Fn() {
-            register(rebuilder(), effects::as_listener()).0
+            register(effects::rebuilder(), effects::as_listener()).0
         }
 
         fn build_counter(CapsuleHandle { mut get, register }: CapsuleHandle) -> usize {
@@ -667,78 +661,76 @@ mod tests {
         assert_eq!(container.read(build_counter), 3);
     }
 
-    /*
-        #[test]
-        fn listener_gets_updates() {
-            use std::sync::{Arc, Mutex};
+    #[test]
+    fn listener_gets_updates() {
+        use std::sync::{Arc, Mutex};
 
-            fn stateful(
-                CapsuleHandle { register, .. }: CapsuleHandle,
-            ) -> (u8, impl CData + Fn(u8)) {
-                let (state, set_state) = register.register(side_effects::state(0));
-                (*state, set_state)
-            }
-
-            let states = Arc::new(Mutex::new(Vec::new()));
-
-            let effect_factory = || ();
-            let listener = {
-                let states = states.clone();
-                move |mut reader: CapsuleReader, _| {
-                    let mut states = states.lock().unwrap();
-                    states.push(reader.get(stateful).0);
-                }
-            };
-
-            let container = Container::new();
-
-            container.read(stateful).1(1);
-            let handle = container.listen(effect_factory, listener.clone());
-            container.read(stateful).1(2);
-            container.read(stateful).1(3);
-
-            drop(handle);
-            container.read(stateful).1(4);
-
-            container.read(stateful).1(5);
-            let handle = container.listen(effect_factory, listener);
-            container.read(stateful).1(6);
-            container.read(stateful).1(7);
-
-            drop(handle);
-            container.read(stateful).1(8);
-
-            let states = states.lock().unwrap();
-            assert_eq!(*states, vec![1, 2, 3, 5, 6, 7]);
+        fn stateful(CapsuleHandle { register, .. }: CapsuleHandle) -> (u8, impl CData + Fn(u8)) {
+            register.register(effects::cloned_state(0))
         }
 
-        #[test]
-        fn listener_side_effects_update() {
-            use std::sync::{Arc, Mutex};
+        let states = Arc::new(Mutex::new(Vec::new()));
 
-            fn rebuildable(
-                CapsuleHandle { register, .. }: CapsuleHandle,
-            ) -> (impl CData + Fn()) {
-                register.register(side_effects::rebuilder())
+        let effect_factory = || ();
+        let listener = {
+            let states = states.clone();
+            move |mut reader: CapsuleReader, _| {
+                let mut states = states.lock().unwrap();
+                states.push(reader.get(stateful).0);
             }
+        };
 
-            let states = Arc::new(Mutex::new(Vec::new()));
+        let container = Container::new();
 
-            let container = Container::new();
-            fn thing() -> impl SideEffect<'a, Api = bool> {
-                side_effects::is_first_build()
-            }
-            let handle = container.listen(thing, |mut get, is_first_build| {
-                get.get(rebuildable);
-                states.clone().lock().unwrap().push(is_first_build);
-            });
+        container.read(stateful).1(1);
+        let handle = container.listen(effect_factory, listener.clone());
+        container.read(stateful).1(2);
+        container.read(stateful).1(3);
 
-            container.read(rebuildable)();
+        drop(handle);
+        container.read(stateful).1(4);
 
-            let states = states.lock().unwrap();
-            assert_eq!(*states, vec![true, false])
+        container.read(stateful).1(5);
+        let handle = container.listen(effect_factory, listener);
+        container.read(stateful).1(6);
+        container.read(stateful).1(7);
+
+        drop(handle);
+        container.read(stateful).1(8);
+
+        let states = states.lock().unwrap();
+        assert_eq!(*states, vec![1, 2, 3, 5, 6, 7]);
+    }
+
+    #[test]
+    fn listener_side_effects_update() {
+        use std::sync::{Arc, Mutex};
+
+        fn rebuildable(CapsuleHandle { register, .. }: CapsuleHandle) -> (impl CData + Fn()) {
+            register.register(effects::rebuilder())
         }
-    */
+
+        let states = Arc::new(Mutex::new(Vec::new()));
+
+        let container = Container::new();
+        let handle = {
+            let states = states.clone();
+            container.listen(
+                || effects::is_first_build(),
+                move |mut get, is_first_build| {
+                    let _ = get.get(rebuildable);
+                    states.lock().unwrap().push(is_first_build);
+                },
+            )
+        };
+
+        container.read(rebuildable)();
+
+        let states = states.lock().unwrap();
+        assert_eq!(*states, vec![true, false]);
+
+        drop(handle);
+    }
 
     #[test]
     fn eq_check_skips_unneeded_rebuilds() {
@@ -762,8 +754,7 @@ mod tests {
 
         fn stateful(CapsuleHandle { register, .. }: CapsuleHandle) -> (u32, impl CData + Fn(u32)) {
             increment_build_count(stateful);
-            let (state, set_state) = register.register(effects::state(0));
-            (*state, set_state)
+            register.register(effects::cloned_state(0))
         }
 
         struct UnchangingIdempotentDep;
@@ -960,8 +951,7 @@ mod tests {
     #[test]
     fn dynamic_and_static_capsules() {
         fn stateful(CapsuleHandle { register, .. }: CapsuleHandle) -> (u8, impl CData + Fn(u8)) {
-            let (state, set_state) = register.register(effects::state(0));
-            (*state, set_state)
+            register.register(effects::cloned_state(0))
         }
         struct Cell(u8);
         impl Capsule for Cell {
@@ -999,8 +989,7 @@ mod tests {
     #[test]
     fn complex_dependency_graph() {
         fn stateful_a(CapsuleHandle { register, .. }: CapsuleHandle) -> (u8, impl CData + Fn(u8)) {
-            let (state, set_state) = register.register(effects::state(0));
-            (*state, set_state)
+            register.register(effects::cloned_state(0))
         }
 
         fn a(CapsuleHandle { mut get, .. }: CapsuleHandle) -> u8 {
@@ -1092,15 +1081,13 @@ mod tests {
         fn two_side_effects_capsule(
             CapsuleHandle { register, .. }: CapsuleHandle,
         ) -> ((u8, impl CData + Fn(u8)), (u8, impl CData + Fn(u8))) {
-            let ((s1, ss1), (s2, ss2)) = register.register((effects::state(0), effects::state(1)));
-            ((*s1, ss1), (*s2, ss2))
+            register.register((effects::cloned_state(0), effects::cloned_state(1)))
         }
 
         fn another_capsule(
             CapsuleHandle { register, .. }: CapsuleHandle,
         ) -> (u8, impl CData + Fn(u8)) {
-            let (state, set_state) = register.register(effects::state(2));
-            (*state, set_state)
+            register.register(effects::cloned_state(2))
         }
 
         fn batch_all_updates_action(

--- a/rearch/src/side_effect_registrar.rs
+++ b/rearch/src/side_effect_registrar.rs
@@ -34,7 +34,7 @@ impl<'a> SideEffectRegistrar<'a> {
     }
 
     /// Registers the given side effect.
-    pub fn register<S: SideEffect<'a>>(self, effect: S) -> S::Api {
+    pub fn register<S: SideEffect>(self, effect: S) -> S::Api<'a> {
         effect.build(self)
     }
 }
@@ -74,8 +74,8 @@ impl<'a> SideEffectRegistrar<'a> {
 
 // One arg register needs its own impl because tuples with one effect don't impl SideEffect
 #[cfg(feature = "better-api")]
-impl<'a, S: SideEffect<'a>> FnOnce<(S,)> for SideEffectRegistrar<'a> {
-    type Output = S::Api;
+impl<'a, S: SideEffect> FnOnce<(S,)> for SideEffectRegistrar<'a> {
+    type Output = S::Api<'a>;
     extern "rust-call" fn call_once(self, (effect,): (S,)) -> Self::Output {
         self.register(effect)
     }
@@ -84,8 +84,8 @@ macro_rules! generate_side_effect_registrar_fn_impl {
     ($($types:ident),*) => {
         #[allow(unused_parens, non_snake_case)]
         #[cfg(feature = "better-api")]
-        impl<'a, $($types: SideEffect<'a>),*> FnOnce<($($types,)*)> for SideEffectRegistrar<'a> {
-            type Output = ($($types::Api),*);
+        impl<'a, $($types: SideEffect),*> FnOnce<($($types,)*)> for SideEffectRegistrar<'a> {
+            type Output = ($($types::Api<'a>),*);
             extern "rust-call" fn call_once(self, args: ($($types),*)) -> Self::Output {
                 self.register(args)
             }

